### DISCUSSION TITLE
The use of poll() inside the multi interface

### DISCUSSION THREAD

**Discussion by HenrikHolst** [curl/curl#8406] [General]
Doing strace on a few of my applications I noticed that curl calls poll() with a 0 timeout inside curl_multi_socket_action if CURL_SOCKET_TIMEOUT is set.

Here is such a strace :

First CURLMOPT_TIMERFUNCTION is called with timeout=0, in this application I use a timerfd to handle this and to simply the code I simply set the timeout to sec=0,nsec=1 so that the timerfd will expire immediately:
timerfd_settime(6, 0, {it_interval={tv_sec=0, tv_nsec=0}, it_value={tv_sec=0, tv_nsec=1}}, NULL) = 0

I use epoll as the event handler and due to the above, epoll responds that we can read from the timerfd:
epoll_pwait(5, [{EPOLLIN, {u32=6, u64=6}}], 2, 1000, [], 8) = 1

Read from the timerfd:
read(6, "\1\0\0\0\0\0\0\0", 8)          = 8

Since read from the timerfd was a success (it signaled that the timer had expired one time) we now call curl_multi_socket_action (curlm, CURL_SOCKET_TIMEOUT, 0, &running_handles); :
poll([{fd=8, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
sendto(8, "GET /content/since/1644333499000"..., 225, MSG_NOSIGNAL, NULL, 0) = 225
poll([{fd=8, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)

And for some reason curl_multi_socket_action first calls poll() with a zero timeout, then it does a successful send() of the whole request (hard to see here but this send is complete so the next action is to read from the connection), but instead of returning to let the users event handler signal for readability immediately, it instead does another call to poll() with a zero timeout and since it gets 0 as response to it curl_multi_socket_action() simply returns.

Now my guess is that curl does the second poll in an attempt to see if it could do a read immediately instead of having to wait for the users event handler, why the first call to poll is done I'm not entire sure but it almost looks like it's there to check if there really was a timeout (which is the whole reason curl_multi_socket_action was called in the first place).

IMHO none of the poll calls should be there since this is already taken care of by the users event handler so they feel completely unnecessary. I'm also curios on what might have happened if the second poll would have signalled readability instead of timeout, would curl have tried to do an internal read by itself and thus kindof breaking objective #1 from the multi interface "1. Enable a "pull" interface. The application that uses libcurl decides where and when to ask libcurl to get/send data. " and could we end up in a theoretical infinite loop inside curl_multi_socket_action here if the server would send say a single byte with perfect timing each time curl calls poll?

In any case this behaviour makes curl perform 3 syscalls instead of just 1, then there might be some good reasoning behind the use of poll here, hence why I opened this as a discussion and not as an issue.

**Comment by bagder** [comment]
Well, unnecessary function calls are unnecessary. I think you're seeing the result of an oversight and/or a shortcoming due to the function being used from multiple different API entry points where `curl_multi_socket_action` is one of them.

**Comment by HenrikHolst** [comment]
Ok, I'll try and see if I can find why they are called and see if they can be eliminated.

**Comment by HenrikHolst** [comment]
So I dug a bit deeper here and now I don't know the history of the code so perhaps things have been moved around over time, but as the code is written today there is a deliberate call to poll() after a write inside the multi interface. This is done in Curl_readwrite() and the only function that ever calls this is multi_runsingle() so perhaps this was used by the easy handle to begin with and then moved over to the multi interface and the call to Curl_socket_check(which is where the poll happens) was left by accident?!

I can remove that call to poll() with a simple patch, I just want to give people time to think over why it was put there in the first place so I don't remove something that was added deliberately. It feels highly unlikely since it's inside the multi interface but still I don't have the experience with the code base that the rest of you have.

Then during connection setup there are two calls to poll() done in series. They are done inside create_conn which first calls prune_dead_connections which in turn calls poll() by way of SocketIsDead() with a zero timeout to check if the socket is open or closed, then create_conn calls ConnectionExists which again calls SocketIsDead() to see if the socket is open or closed.

So the same socket is checked for open/close twice during the connection setup, at least one of them feels unnecessary while one of them might be unavoidable since curl needs to do some check to see if the socket is alive or not and doing that without a syscall might be impossible.

**Comment by HenrikHolst** [comment]
Strangely enough is that removing the call to poll() with a zero timeout in Curl_readwrite() makes the unit checks fail immediately with :

RUN: Unknown server on our http port: 43451 (28)
RUN: HTTP server failed verification

So either the multi interface really needs to call poll (which I highly doubt and I have tested the patched curl with my own multi-interface applications) or there are some parts of the test server or unit test framework that requires it.

**Comment by bagder** [comment]
I propose you make a draft PR with the patch and then I can have a look in my end to see if I can figure out what or why that test error happens.

**Comment by HenrikHolst** [comment]
Done so. Regarding the other calls to poll() it turns out that prune_dead_connections() only does this every 1000ms so the overhead here should be negligible. I just happened to see if since my particular test run once every second so I happened to trigger both checks every single time.

**Comment by bagder** [comment]
> it turns out that prune_dead_connections() only does this every 1000 ms

`prune_dead_connections()` is only called when *creating a new connection* and when so, not more frequently than once per 1000ms.

**Comment by HenrikHolst** [comment]
Depends on how you define "creating new connection", it's called each time you initiate a transfer. For my example I fetch the same URL from the same server and the exact same connection is reused since curl finds it in the connection cache but prune_dead_connections() is called every single time (since I do fetch, sleep 1000, fetch, ...) but since it's only called each 1000ms the overhead of the poll() syscall is negligible.

ConnectionExists is "worse" since it's calling poll() for every single transfer, but it's probably needed to so not accidentally reuse a dead connection.
